File: platform/execution-impl/src/com/intellij/terminal/CompositeFilterWrapper.java
// Copyright 2000-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.intellij.terminal;

import com.intellij.execution.filters.CompositeFilter;
import com.intellij.execution.filters.ConsoleFilterProvider;
import com.intellij.execution.filters.Filter;
import com.intellij.execution.impl.ConsoleViewUtil;
import com.intellij.openapi.Disposable;
import com.intellij.openapi.project.Project;
import com.intellij.psi.search.GlobalSearchScope;
import com.intellij.util.concurrency.annotations.RequiresReadLock;
import com.intellij.util.containers.ContainerUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collections;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

final class CompositeFilterWrapper {

  private final Project myProject;
  private final TerminalExecutionConsole myConsole;
  private final List<Filter> myCustomFilters = new CopyOnWriteArrayList<>();
  private volatile CompositeFilter myCompositeFilter;

  CompositeFilterWrapper(@NotNull Project project, @Nullable TerminalExecutionConsole console, @NotNull Disposable disposable) {
    myProject = project;
    myConsole = console;
    ConsoleFilterProvider.FILTER_PROVIDERS.addChangeListener(() -> {
      myCompositeFilter = null;
    }, disposable);
  }

  void addFilter(@NotNull Filter filter) {
    myCustomFilters.add(filter);
    myCompositeFilter = null;
  }

  @RequiresReadLock
  private @NotNull List<Filter> computeConsoleFilters() {
    if (myProject.isDefault()) {
      return Collections.emptyList();
    }
    if (myProject.isDisposed()) {
      return Collections.emptyList();
    }
    return ConsoleViewUtil.computeConsoleFilters(myProject, myConsole, GlobalSearchScope.allScope(myProject));
  }

  @RequiresReadLock
  @NotNull CompositeFilter getCompositeFilter() {
    CompositeFilter filter = myCompositeFilter;
    if (filter != null) {
      return filter;
    }
    List<Filter> predefinedFilters = computeConsoleFilters();
    filter = new CompositeFilter(myProject, ContainerUtil.concat(myCustomFilters, predefinedFilters));
    filter.setForceUseAllFilters(true);
    myCompositeFilter = filter;
    return filter;
  }
}

File: plugins/terminal/src/org/jetbrains/plugins/terminal/exp/CompositeFilterWrapper.kt
// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.plugins.terminal.exp

import com.intellij.execution.filters.CompositeFilter
import com.intellij.execution.filters.ConsoleFilterProvider
import com.intellij.execution.filters.Filter
import com.intellij.execution.impl.ConsoleViewUtil
import com.intellij.openapi.Disposable
import com.intellij.openapi.application.runReadAction
import com.intellij.openapi.project.Project
import com.intellij.psi.search.GlobalSearchScope

internal class CompositeFilterWrapper(private val project: Project, disposable: Disposable) {

  @Volatile
  private var cachedCompositeFilter: CompositeFilter? = null

  init {
    ConsoleFilterProvider.FILTER_PROVIDERS.addChangeListener({ cachedCompositeFilter = null }, disposable)
  }

  private fun createCompositeFilters(): List<Filter> {
    if (project.isDefault) {
      return emptyList()
    }
    return runReadAction {
      if (project.isDisposed) {
        return@runReadAction emptyList<Filter>()
      }
      ConsoleViewUtil.computeConsoleFilters(project, null, GlobalSearchScope.allScope(project))
    }
  }

  val compositeFilter: CompositeFilter
    get() {
      cachedCompositeFilter?.let {
        return it
      }
      val resultCompositeFilter = CompositeFilter(project, createCompositeFilters()).also {
        it.setForceUseAllFilters(true)
      }
      cachedCompositeFilter = resultCompositeFilter
      return resultCompositeFilter
    }
}
